### PR TITLE
8262368: wrong verifier message for bogus return type

### PR DESCRIPTION
**Pull request by hseigel**
Please review this small change to fix some verifier error messages.  If a method's descriptor has a void return signature, but contains a bytecode such as 'ireturn' or 'areturn'. then it will get a VerifyError exception with the message "Method descriptor has a void return type".   Conversely, if a method's descriptor has a non-void return signature, but contains a 'return' bytecode then it will get a VerifyError exception with the message "Method descriptor has a non-void return type".

A sample error message looks like:
Error: Unable to initialize main class Hi
Caused by: java.lang.VerifyError: Method descriptor has a void return type
Exception Details:
  Location:
    Hi.noRet(I)V @1: ireturn
  Reason:
    Type integer (current frame, stack[0]) is not assignable to top (from method signature)
  Current Frame:
    bci: @1
    flags: { }
    locals: { integer }
    stack: { integer }
  Bytecode:
    0000000: 08ac

The fix was tested with Mach5 tiers 1 and 2 on Linux, Mac OS, and Windows, tiers 3-5 on Linux x64. and the JCK Lang and VM tests on Linux x64.

Thanks, Harold

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8262368](https://bugs.openjdk.java.net/browse/JDK-8262368): wrong verifier message for bogus return type

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2757/head:pull/2757`
`$ git checkout pull/2757`

### CODE REVIEW

File: src/hotspot/share/classfile/verifier.cpp [hunk 1]
```diff
@@ -1675,7 +1675,7 @@ void ClassVerifier::verify_method(const methodHandle& m, TRAPS) {
         case Bytecodes::_return :
           if (return_type != VerificationType::bogus_type()) {
             verify_error(ErrorContext::bad_code(bci),
-                         "Method expects a return value");
+                         "Method descriptor has a non-void return type");
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/verifier.cpp:1678]: I prefer the original text. The new text makes me have to think about what the actual problem is too much.

File: src/hotspot/share/classfile/verifier.cpp [hunk 2]
```diff
@@ -3149,7 +3149,7 @@ void ClassVerifier::verify_return_value(
   if (return_type == VerificationType::bogus_type()) {
     verify_error(ErrorContext::bad_type(bci,
         current_frame->stack_top_ctx(), TypeOrigin::signature(return_type)),
-        "Method expects a return value");
+        "Method descriptor has a void return type");
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/verifier.cpp:3152]: I would prefer "Method does not expect a return value".

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Hi Harold,

Sorry I don't like the new messages. See below.

Thanks,
David

File: src/hotspot/share/classfile/verifier.cpp (hunk 2 above)
Comment (hseigel) [reply src/hotspot/share/classfile/verifier.cpp]: I changed the messages to what you suggested.
Thanks for reviewing this change.
Harold

File: test/hotspot/jtreg/runtime/verifier/ReturnMsgs.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8262368
+ * @summary Test that VerifyError messages are correct when return bytecodes
+ *          andd method signatures do not match.
```
Comment (coleenp) [inline test/hotspot/jtreg/runtime/verifier/ReturnMsgs.java:28]: typo, andd

**Review by coleenp — APPROVED** [review]
Looks good with the fixed message.  I agree that it took me an extra minute to understand the new messages, so the old ones (corrected) are fine.  Nice simple test.

File: src/hotspot/share/classfile/verifier.cpp [hunk 4]
```diff
@@ -3149,7 +3149,7 @@ void ClassVerifier::verify_return_value(
   if (return_type == VerificationType::bogus_type()) {
     verify_error(ErrorContext::bad_type(bci,
         current_frame->stack_top_ctx(), TypeOrigin::signature(return_type)),
-        "Method expects a return value");
+        "Method does not expect a return value");
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/verifier.cpp:3152]: You've changed the wrong line. This one does expect a return value

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
You fixed the wrong one.

Cheers,
David

File: src/hotspot/share/classfile/verifier.cpp (hunk 4 above)
Comment (hseigel) [reply src/hotspot/share/classfile/verifier.cpp:3152]: Hi David,
If a method's descriptor has a void return type, but the method contains a bytecode such as areturn, then I think the message "Method does not expect a return type" is correct.  Perhaps the word 'expect' is ambiguous?  If so, then I think a more explicit message, similar to my original suggestion, should be used.
Harold

File: src/hotspot/share/classfile/verifier.cpp [hunk 5]
```diff
@@ -3149,7 +3149,7 @@ void ClassVerifier::verify_return_value(
   if (return_type == VerificationType::bogus_type()) {
```
Comment (coleenp) [inline src/hotspot/share/classfile/verifier.cpp:3149]: return_type is the type expected from the function, which here is void (I guess void == bogus_type, but ok).
So the message is correct.

**Comment by coleenp** [comment]
Let's use javac to help us with the error message:
$ more X.java
class X {
void foo() { return obj; }
int bar() { return; }
}
$ javac X.java
X.java:2: error: incompatible types: unexpected return value
void foo() { return obj; } <===  This is line 3152
                    ^
X.java:3: error: incompatible types: missing return value
int bar() { return; }       <=== This is line 1678
            ^
2 errors

File: src/hotspot/share/classfile/verifier.cpp (hunk 4 above)
Comment (dholmes-ora) [reply src/hotspot/share/classfile/verifier.cpp:3152]: Okay I see now the original bug report was incorrect and the referenced code is correct. But the above code is actually incorrect and now fixed.

**Review by dholmes-ora — APPROVED** [review]
Thanks,
David

**Comment by hseigel** [comment]
Thanks Coleen and David for the reviews!

/integrate

**Pull request closed without merging**